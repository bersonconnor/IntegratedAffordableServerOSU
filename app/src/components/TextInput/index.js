import React, { useState } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import "./scss/text-input.scss";

const TextInput = ({
  containerClassName,
  handleChangeByParent,
  id,
  labelText,
  password,
  parentValue
}) => {
  const [hasValue, setHasValue] = useState(parentValue ? true : false);
  const [hasFocus, setHasFocus] = useState(false);
  const [componentValue, setComponentValue] = useState("");

  const handleFocus = () => {
    setHasFocus(true);
  };

  const handleBlur = () => {
    setHasFocus(false);
  };

  const handleChange = event => {
    parentValue !== undefined
      ? handleChangeByParent(event)
      : setComponentValue(event.target.value);

    event.target.value ? setHasValue(true) : setHasValue(false);
  };

  return (
    <div className={classNames("text-input__container", containerClassName)}>
      <input
        className={classNames("text-input", {
          "text-input--has-value": hasValue
        })}
        onBlur={handleBlur}
        onChange={handleChange}
        onFocus={handleFocus}
        id={id}
        type={password ? "password" : "text"}
        value={parentValue || componentValue}
      />
      <label
        htmlFor={id}
        className={classNames("text-input__label", {
          "text-input__label--inline": !hasValue && !hasFocus,
          "text-input__label--float": hasFocus || hasValue,
          "text-input__label--float--focused": hasFocus
        })}
      >
        {labelText}
      </label>
    </div>
  );
};

TextInput.propTypes = {
  /**
   * Class to be applied to the container of TextInput.
   * Typically used for resizing TextInput
   */
  containerClassName: PropTypes.string,
  /**
   * Logic that handles change event if state is kept
   * in parent node
   */
  handleChangeByParent: PropTypes.func,
  /**
   * ID of the TextInput
   * (NOT INTENDED FOR STYLING)
   */
  id: PropTypes.string.isRequired,
  /**
   * Text shown by the label
   */
  labelText: PropTypes.string.isRequired,
  /**
   * If `true`, text typed in by user will be hidden
   */
  password: PropTypes.bool,
  /**
   * Value passed in if state is handled by parent
   * node
   */
  parentValue: PropTypes.string
};

TextInput.defaultProps = {
  containerClassName: undefined,
  handleChangeByParent: () => {},
  password: false,
  parentValue: undefined
};

export default TextInput;
