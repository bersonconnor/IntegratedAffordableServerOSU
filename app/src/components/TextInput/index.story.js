import React from "react";
import { storiesOf } from "@storybook/react";
import TextInput from "./index";

storiesOf("TextInput", module)
  .add("default", () => <TextInput labelText="Username" id="username" />)
  .add("controlled", () => (
    <TextInput
      labelText="Username"
      id="username"
      parentValue="VALUE SHOULD NOT CHANGE"
    />
  ));
