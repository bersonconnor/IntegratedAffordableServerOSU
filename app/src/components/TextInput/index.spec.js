import React, { useState } from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import TextInput from "./";

const testLabelText = "Test";
const testID = "test";

const createTextInput = (props = {}) => (
  <TextInput labelText={testLabelText} id={testID} {...props} />
);

describe("TextInput", () => {
  describe("Snapshots", () => {
    it("should match snapshot for default TextInput", () => {
      expect(renderer.create(createTextInput()).toJSON()).toMatchSnapshot();
    });

    it("should match snapshot for password TextInput", () => {
      expect(
        renderer.create(createTextInput({ password: true })).toJSON()
      ).toMatchSnapshot();
    });
  });

  describe("Controlled vs Uncontrolled", () => {
    const testString = "Test String";

    it("should handle its own state when no parentValue passed in - uncontrolled", () => {
      const wrapper = shallow(createTextInput());
      wrapper
        .find("input")
        .simulate("change", { target: { value: testString } });
      expect(wrapper.find("input")).toHaveValue(testString);
    });

    it("should not change value when parentValue passed in - controlled", () => {
      const wrapper = shallow(createTextInput({ parentValue: testString }));
      wrapper.find("input").simulate("change", { target: { value: "EEP" } });
      expect(wrapper.find("input")).toHaveValue(testString);
    });

    it("should change parent's state when handleChangeByParent passed in - controlled", () => {
      const ParentComponent = () => {
        const [value, setValue] = useState("");

        const handleChange = event => {
          setValue(event.target.value);
        };

        return (
          <div>
            {createTextInput({
              parentValue: value,
              handleChangeByParent: handleChange
            })}
            <p>{value}</p>
          </div>
        );
      };

      const wrapper = shallow(<ParentComponent />);

      wrapper
        .find(TextInput)
        .dive()
        .find("input")
        .simulate("change", { target: { value: testString } });

      expect(wrapper.find("p")).toHaveText(testString);
    });
  });
});
