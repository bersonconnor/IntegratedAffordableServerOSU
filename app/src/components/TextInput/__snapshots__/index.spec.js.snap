// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Snapshots should match snapshot for default TextInput 1`] = `
<div
  className="text-input__container"
>
  <input
    className="text-input"
    id="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <label
    className="text-input__label text-input__label--inline"
    htmlFor="test"
  >
    Test
  </label>
</div>
`;

exports[`TextInput Snapshots should match snapshot for password TextInput 1`] = `
<div
  className="text-input__container"
>
  <input
    className="text-input"
    id="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="password"
    value=""
  />
  <label
    className="text-input__label text-input__label--inline"
    htmlFor="test"
  >
    Test
  </label>
</div>
`;
